Scripts/Source/UD_RandomRestraintManager.psc

```diff
@@ -162,51 +181,87 @@ Armor Function getRandomDeviceByKeyword(Actor akActor,Keyword kwKeyword)
+                LL = zadDL.zad_dev_heavyrestraints
             endif
         endif
     elseif kwKeyword == libs.zad_DeviousSuit
-        if !akActor.wornhaskeyword(libs.zad_DeviousSuit)
-            res = getRandomFormFromLeveledlist(zadDL.zad_dev_suits_catsuits);getRandomFormFromFormlist(UD_RandomDeviceList_Suit)
+        if !akActor.wornhaskeyword(libs.zad_DeviousHeavyBondage)
+            LL = zadDL.zad_dev_suits                                ; can use any suit, including straitjackets
+        else
+            int rnd_suit = Utility.randomInt(0,4)                   ; can solve this by adding leveled list without SJs...
+            if rnd_suit == 0                                        ; but this should be good enough
+                LL = zadDL.zad_dev_suits_catsuits
+            elseif rnd_suit == 1
+                LL = zadDL.zad_dev_suits_hobbledresses
+            elseif rnd_suit == 2
+                LL = zadDL.zad_dev_suits_relaxed_hobbledresses
+            elseif rnd_suit == 3
+                LL = zadDL.zad_dev_suits_hobbledresses
+            elseif rnd_suit == 4
+                LL = zadDL.zad_dev_suits_formaldresses
+            endif
         endif
     elseif kwKeyword == libs.zad_DeviousPlugVaginal
-        res = getRandomFormFromLeveledlist(zadDL.zad_dev_plugs_vaginal);getRandomFormFromFormlist(UD_RandomDeviceList_PlugVaginal)
+        LL = zadDL.zad_dev_plugs_vaginal
     elseif kwKeyword == libs.zad_DeviousPlugAnal
-        res = getRandomFormFromLeveledlist(zadDL.zad_dev_plugs_anal);getRandomFormFromFormlist(UD_RandomDeviceList_PlugAnal)
+        LL = zadDL.zad_dev_plugs_anal
     elseif kwKeyword == libs.zad_DeviousPiercingsVaginal
-        res = getRandomFormFromLeveledlist(zadDL.zad_dev_piercings_vaginal);getRandomFormFromFormlist(UD_RandomDeviceList_PiercingVag)
+        LL = zadDL.zad_dev_piercings_vaginal
     elseif kwKeyword == libs.zad_DeviousPiercingsNipple
-        res = getRandomFormFromLeveledlist(zadDL.zad_dev_piercings_nipple);getRandomFormFromFormlist(UD_RandomDeviceList_PiercingNipple)
+        LL = zadDL.zad_dev_piercings_nipple
     elseif kwKeyword == libs.zad_DeviousGloves
-        res = getRandomFormFromLeveledlist(zadDL.zad_dev_gloves);getRandomFormFromFormlist(UD_RandomDeviceList_Gloves)
+        LL = zadDL.zad_dev_gloves
     elseif kwKeyword == libs.zad_DeviousHood
-        res = getRandomFormFromLeveledlist(zadDL.zad_dev_hoods);getRandomFormFromFormlist(UD_RandomDeviceList_Hood)
+        LL = zadDL.zad_dev_hoods
+    elseif kwKeyword == libs.zad_DeviousCorset && !akActor.wornhaskeyword(libs.zad_DeviousHarness)
+        LL = zadDL.zad_dev_corsets
+    elseif kwKeyword == libs.zad_DeviousHarness && !akActor.wornhaskeyword(libs.zad_DeviousCorset)
+        LL = zadDL.zad_dev_harnesses
+    else
+        return none
+    endif
+
+    if LL
+        int tries = 10                        ; 10 attempts
```

Comment:
Reduce this number to ~2. If it failed first time, there is big chance it will fail again on next attempt. So forcing this check 10 times can freeze player for few minutes before all conflicts are resolved. You should also add some console trace, so users can see which devices are conflicting. In this case, I think the best option is UDmain.Warning or just GWarning

Reply:
Actually, my testing shown that this reroll works really fast!
It's the most useful when pulling various harnesses, some can have collars, some can have chastity.
So it tries several times to pull variety of device which is not in conflict with others.

Maybe leave it as it is for testing, and if some really nasty hold-ups arise, then reduce?

Reply:
I have tested it and it took very long time for all devices to be locked. But it might have been caused by something else. Will have to investigate further later. For now, I will ignore this point. After merge, I will have to do some more benchmark testing.